torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,609 @@
-    if (!parameter.grad().defined()) {
-      views.push_back(parameter.new_empty({parameter.numel()}).zero_());
+  for (const auto& p : param_groups_.at(0).params()) {
+    if (!p.grad().defined()) {
+      views.emplace_back(p.new_empty({p.numel()}).zero_());
     }
-    else if (parameter.grad().is_sparse()) {
-      views.push_back(parameter.grad().to_dense().view(-1));
+    else if (p.grad().is_sparse()) {
+      views.emplace_back(p.grad().to_dense().view(-1));
     }
     else {
-      views.push_back(parameter.grad().view(-1));
+      views.emplace_back(p.grad().view(-1));
+    }
+  }
+  return torch::cat(views, 0);
+}
+
+int64_t LBFGS::_numel() {
+  if (_numel_cache == c10::nullopt) {
+    auto res = 0;
+    for (const auto& p : param_groups_.at(0).params()) {
+      res += p.numel();
     }
+    _numel_cache = res;
   }
-  return torch::cat(views);
+  return *_numel_cache;
 }
 
-void LBFGS::add_grad(const torch::Tensor& step_size, const Tensor& update) {
-  NoGradGuard guard;
-  int64_t offset = 0;
-  for (auto& parameter : parameters_) {
-    int64_t numel = parameter.numel();
-    parameter.add_(
-        update.slice(0, offset, offset + numel, 1).view_as(parameter),
-        step_size.item<float>());
+void LBFGS::_add_grad(const double step_size, const Tensor& update) {
+  auto offset = 0;
+  for (auto& p : param_groups_.at(0).params()) {
+    auto numel = p.numel();
+    // view as to avoid deprecated pointwise semantics
+    p.add_(update.index(
+      {at::indexing::Slice(offset, offset + numel)}).view_as(p), step_size);
     offset += numel;
   }
+  TORCH_INTERNAL_ASSERT(offset == _numel());
+}
+
+void LBFGS::_set_param(const std::vector<Tensor>& params_data) {
+  auto& _params = param_groups_.at(0).params();
+  TORCH_INTERNAL_ASSERT(params_data.size() == _params.size());
+  for (size_t i = 0; i < _params.size(); i++) {
+    _params.at(i).copy_(params_data.at(i));
+  }
+}
+
+std::vector<Tensor> LBFGS::_clone_param() {
+  std::vector<Tensor> result;
+  for (const auto& p : param_groups_.at(0).params()) {
+    result.emplace_back(p.clone(at::MemoryFormat::Contiguous));
+  }
+  return result;
+}
+
+std::tuple<double, Tensor> LBFGS::_directional_evaluate(
+  const LossClosure& closure, const std::vector<Tensor>& x, double t, const Tensor& d) {
+    _add_grad(t, d);
+    double loss;
+    {
+      torch::AutoGradMode enable_grad(true);
+      loss = closure().item<double>();
+    }
+    auto flat_grad = _gather_flat_grad();
+    _set_param(x);
+    return std::make_tuple(loss, flat_grad);
+}
+
+double _cubic_interpolate(
+  double x1, double f1, double g1, double x2, double f2, double g2,
+  c10::optional<std::tuple<double, double>> bounds = c10::nullopt) {
+  // ported from https://github.com/torch/optim/blob/master/polyinterp.lua
+  // Compute bounds of interpolation area
+  double xmin_bound, xmax_bound;
+  if (bounds != c10::nullopt) {
+    xmin_bound = std::get<0>(*bounds);
+    xmax_bound = std::get<1>(*bounds);
```

Comment:
might be easier to write
```suggestion
    std::tie(xmin_bound, xmax_bound) = *bounds;
```